### PR TITLE
[CS2113T-F11-3] KAJI

### PR DESCRIPTION
**Pull request by Jane-Ng**
KAJI is a schedule manager that implements Spaced Repetition for university students, optimised for use via a Command Line Interface (CLI).

### CODE REVIEW

**Comment by PraveenElango** [comment]
DG Comment 1

LogicSequenceDiagram

In the above UML diagram (under 3.2 Logic Component), would it be better if `c` is specified explicitly?

**Comment by PraveenElango** [comment]
DG Comment 2 

removechap_seq_diagram

In the above UML diagram (4.2.4. Remove Chapter Feature), should we consider removing the exception shown in the diagram? A sequence diagram aims to show how objects interact and not to fully specify all the possible behaviors such as errors and exceptions.

**Comment by PraveenElango** [comment]
DG Comment 3 

removechap_seq_diagram

In the above UML diagram (4.2.4. Remove Chapter Feature), should we specify what is being ultimately returned by the RemoveChapterCommand?

**Comment by PraveenElango** [comment]
DG Comment 4 

In section 4.4.2. Scheduling The Chapters Feature, while it is comprehensive to explain each step of the process on how the chapters are scheduled and the associated attributes such as `LocalDate`, `deckInterval`, `previousInterval`, would it possibly be too wordy? Should we consider using a UML sequence diagam to illustrate the sequence of events so as to make it clearer and also make the explanation more concise?

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -1,34 +1,938 @@
...
+* `Paser`: Parses user input into specific command type.
+* `Scheduler`: Schedules the revision schedule.
+* `Storage`: Reads data from, and writes data to, the hard disk.
+
+### 3.1. Ui Component 
+(Jia Ern)
+
+![Class Diagram of Ui Component](DG_Images/ui_component.PNG)
+
+The UI component consists of a main class — Ui.java. The Kaji, Logic and Storage components have a dependency on the UI component due to the need to take in user input and show the results of execution.
```
Comment (slightlyharp) [inline docs/DeveloperGuide.md]: image
should the line from kaiji to Ui be dashed line?

Comment (slightlyharp) [inline docs/DeveloperGuide.md]: Should there be a description for the Model Class?

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -1,34 +1,938 @@
...
+![](images/ExcludeCommand.png)
+
+![](images/addModuleToExclusion.png)
+
+![](images/addChapterToExclusion.png)
+
+![](images/removeModuleFromExclusion.png)
+
+![](images/removeChapterFromExclusion.png)
+<br>
```
Comment (luziyi9898) [inline docs/DeveloperGuide.md]: Perhaps the explanation/visuals used here is a little repetitive? Consider further splitting it into smaller sub-sections, or shorten the explanation when explaining the subsequent commands that are highly similar to previous ones.

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -1,34 +1,938 @@
...
+3. The command execution can affect the `Model` and `Storage` (e.g. adding a module).
+4. The `Scheduler` is used in some command execution to schedule the due date of a chapter.
+5. The result of the command execution is passed to the `Ui`.
+
+Given below is the Sequence Diagram for interactions within the `Logic` component for the `parse("edit 1 CS2113T")` API call:
+
+<p align="center">
+  <img src="DG_Images/LogicSequenceDiagram.png" width="600" alt="Logic Sequence Diagram"/>
+  <br/>Figure <>. Sequence diagram of Logic component  
+</p>
```
Comment (luziyi9898) [inline docs/DeveloperGuide.md]: Maybe you should standardize the way you add diagrams? Right now some of them are aligned and properly labelled with captions while others are not.

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -1,34 +1,938 @@
...
+* Printing error messages when certain exceptions occur. 
+
+### 3.2. Logic Component 
+(Jane)
+
+The Logic component consists of the `Parser`, `Command` and `Scheduler` classes.
+
+<p align="center">
+  <img src="DG_Images/LogicClassDiagram.png" width="600" alt="Logic Class Diagram"/>
+  <br/>Figure <>. Class diagram of Logic component
```
Comment (slightlyharp) [inline docs/DeveloperGuide.md]: image
Would a UML class diagram make it clearer in explaining the relationship?

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -1,34 +1,938 @@
...
+Step 6: A success message of completeing the revision will be shown to the user through `Ui#showToUser()`.
+
+Step 7: `Scheduler#computeDeckDeadline()` then calculates the new deadline for the `chapter` and passes the result to `Chapter#setueBy()` to set the new deadline for the `chapter`.
+
+Step 6: `ReviseCommand#repeatRevision` then repeats the revision session on cards which the user could not answer.
+
+Step 7: Lastly, `ReviseCommand#execute()` calls `HistoryCommand#addHistory()` to keep a record of the chapter revised so that the user can review what revision has been completed.
+
+The following sequence diagram shows how the revise feature works:
+![Sequence Diagram of Revise](DG_Images/revise_seq_diagram.png)
```
Comment (luziyi9898) [inline docs/DeveloperGuide.md]: The diagram used is rather complicated and quite difficult to read given its small font and compact information. Would it be better if you consider splitting it into different diagrams, or omit some of the less important details?

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -1,34 +1,938 @@
...
+
+`GoModuleCommand` class facilitates the proposed access module level feature. It extends `GoCommand` class. 
+
+It implements the following operations:
+* `GoModuleCommand#execute()` -- Inherites from the abstract `GoCommand`. It is overrided to calls `GoModuleCommand#goModule()` and print message to user.
+* `GoModuleCommand#goModule()` -- Accesses a new module level from admin level based on the module name entered by user. If the module name does not match with any of the existing module, it returns a error message. Otherwise, this operation returns a successful message to the `GoModuleCommand#execute` operation.
+
+Shown as the class diagram below, with the inheritance of `GoCommand` and `Command`, `Kaji` is able to execute the operation `GoModuleCommand#execute()` directly. 
+
+![Class Diagram of go module command](UML/GoModuleCommandClass.png)
```
Comment (luziyi9898) [inline docs/DeveloperGuide.md]: image
The diagram is not displaying correctly when I look at it. Would be good if you can check and update it!

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -1,34 +1,938 @@
...
+* `ListModulesCommand#listModules()` — gets and lists all modules in admin level.
+
+For instance, the user wants to list all modules available in `admin`, a detailed description of what happens is shown below:
+
+* Step 1: The user is currently in `admin` level. 
+
+* Step 2: The user enters `list` command to list all modules in `admin` level. 
+
+The following sequence diagram shows how the list modules feature works:
+![Sequence Diagram of List Modules](UML/listmod_seq_diagram.png)
```
Comment (slightlyharp) [inline docs/DeveloperGuide.md]: image
Should there be activation bar in the diagrram?

**Review by luziyi9898 — COMMENTED** [review]
Very nice DG that paid a lot of attentions to the details, and almost all implementations and explanations are supported by relevant diagrams!

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -1,34 +1,938 @@
...
+* `HistoryCommand#listHistory()` — lists the revision completed in the session/in a day.
+
+Given below is an example usage scenario and how the history mechanism behaves at each step:
+
+Step 1: The user launches the application and is currently in the admin level. 
+
+Step 2: The user executes `history` command to load and list the revision completed in the session/in a day.
+
+The following sequence diagram shows how the list chapters feature works:
+![Sequence Diagram of List History](UML/listhistory_seq_diagram.png)
```
Comment (slightlyharp) [inline docs/DeveloperGuide.md]: image
Should the arrow point to the previous activation bar instead?

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -1,34 +1,938 @@
...
+5. Locate the `build.gradle` file (not the root folder as you would do in a normal importing) and select it. Click `OK`.
+If asked, choose to `Open as Project` (not `Open as File`).
+7. Click `OK` to accept the default settings
+
+--------------------------------------------------------------------------------------------------------------------
+
+## 3. Design (Architecture)
+(Zeyu)
+
+![Architecture Diagram of Design Component](UML/Architecture.png)
```
Comment (R-Ramana) [inline docs/DeveloperGuide.md]: I think it might be better if your diagrams were standardized as to how they were added into the document. Some of them use the HTML notation, aligned and include caption, the rest just follows a simple markdown format. That is not so much of an issue, but maybe the alignment and the use of caption can be standardized?

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -1,34 +1,938 @@
...
+The main class is `Kaji.java`. It is responsible for:
+
+* Initializing the components in the correct sequence, and connects them up with each other at app launch.
+* Shutting down the components and invoking cleanup methods where necessary at exit.
+
+The rest of the App consists of 8 components:
+* `Access`: Keeps track of the access level of the user.
+* `UI`: The UI of the App.
+* `Command`: Executes the different command types.
+* `Common`: Contains common classes.
```
Comment (R-Ramana) [inline docs/DeveloperGuide.md]: From the diagram and subsequent explanation, I am not sure I fully understand which of the following components are connected to the `Common` component. The `Common` component seems to be isolated from the other components.

File: docs/DeveloperGuide.md (hunk 9 above)
Comment (R-Ramana) [reply docs/DeveloperGuide.md]: Maybe can consider labelling the circled component and state that that is a human-editable output file? It wasn't immediately clear for me, and I had to think of what would come out from Storage. It helped that almost all of us had similar architecture for our projects, may not be for a third-party developer.
Screenshot_1

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -1,34 +1,938 @@
...
+* Printing result of execution for the different commands. 
+* Printing error messages when certain exceptions occur. 
+
+### 3.2. Logic Component 
+(Jane)
+
+The Logic component consists of the `Parser`, `Command` and `Scheduler` classes.
+
+<p align="center">
+  <img src="DG_Images/LogicClassDiagram.png" width="600" alt="Logic Class Diagram"/>
```
Comment (R-Ramana) [inline docs/DeveloperGuide.md]: Maybe in your diagram, you can consider having labels for the other lines since you have already had them for some of the lines.
Screenshot_2

File: docs/DeveloperGuide.md [hunk 12]
```diff
@@ -1,34 +1,938 @@
...
+5. The result of the command execution is passed to the `Ui`.
+
+Given below is the Sequence Diagram for interactions within the `Logic` component for the `parse("edit 1 CS2113T")` API call:
+
+<p align="center">
+  <img src="DG_Images/LogicSequenceDiagram.png" width="600" alt="Logic Sequence Diagram"/>
+  <br/>Figure <>. Sequence diagram of Logic component  
+</p>
+
+>:information_source: <b>Note:</b> The lifeline for `Parser` should end at the destroy marker (X) but due to a limitation of PlantUML, the lifeline reaches the end of diagram.
```
Comment (R-Ramana) [inline docs/DeveloperGuide.md]: If the information source was meant to be an emoji, I can't view it on the webpage.
image

File: docs/DeveloperGuide.md [hunk 13]
```diff
@@ -1,34 +1,938 @@
...
+#### 4.1.2. List Modules Feature
+(Zeyu)
+
+The list modules feature allows the user to list all modules in admin level 
+
+The list modules mechanism is facilitated by `ListModulesCommand`. It extends from the abstract class `ListCommand`. 
+
+In addition, it implements the following operations:
+* `ListModulesCommand#execute()` — oversees entire execution for listing modules.
+* `ListModulesCommand#listModules()` — gets and lists all modules in admin level.
```
Comment (R-Ramana) [inline docs/DeveloperGuide.md:299]: Were these meant to be like `ListModulesCommand.execute()`?

File: docs/DeveloperGuide.md (hunk 7 above)
Comment (R-Ramana) [inline docs/DeveloperGuide.md]: Are the deletions correct? I am a little confused if they actually get deleted.

File: docs/DeveloperGuide.md [hunk 14]
```diff
@@ -1,34 +1,938 @@
...
+* Implements Spaced Repetition for the user 
+
+### 5.2. User Stories
 
 |Version| As a ... | I want to ... | So that I can ...|
 |--------|----------|---------------|------------------|
-|v1.0|new user|see usage instructions|refer to them when I forget how to use the application|
-|v2.0|user|find a to-do item by name|locate a to-do without having to go through the entire list|
+|v1.0|user|revise the flashcards by chapter|do my revision|
+|v1.0|user|remove modules/chapters/flashcard|remove modules/chapters/flashcards that I no longer need from the scheduler|
```
Comment (R-Ramana) [inline docs/DeveloperGuide.md]: Maybe can add more user stories?

File: docs/DeveloperGuide.md [hunk 15]
```diff
@@ -1,34 +1,1015 @@
...
+* Initializing the components in the correct sequence, and connects them up with each other at app launch.
+* Shutting down the components and invoking cleanup methods where necessary at exit.
+
+The rest of the App consists of 8 components:
+* `Access`: Keeps track of the access level of the user.
+* `UI`: The UI of the App.
+* `Command`: Executes the different command types.
+* `Common`: Contains common classes.
+* `Manager`: Holds the data of the App in memory.
+* `Paser`: Parses user input into specific command type.
```
Comment (madbeez) [inline docs/DeveloperGuide.md]: Did you mean to write parser instead of paser?

File: docs/DeveloperGuide.md [hunk 16]
```diff
@@ -1,34 +1,1015 @@
...
+* stores a DueChapter object that manages due chapter data.
+* contains an Admin class that can be created and accessed by Access object. The Admin class contains a ModuleList object to manage data for modules.
+* contains a Module class that can be created and accessed by Access object. The Module class contains a ChapterList object to manage data for chapters.
+* contains a Chapter class that can be created and accessed by Access object. The Chapter class contains a CardList object to manage data for cards.
+* contains a Card object that can be created and accessed by CardList. A Card object represents a flashcard with question, answer and master level of revision.
+
+<p align="center">
+  <img src="UML/model.png" width="800" alt="Architecture Diagram of Model"/>
+  <br/>Figure <>. Architecture Diagram of Model
+</p>
```
Comment (madbeez) [inline docs/DeveloperGuide.md]: To add on to what the others have said, you could consider numbering your diagrams?

File: docs/DeveloperGuide.md [hunk 17]
```diff
@@ -1,34 +1,1015 @@
...
+* `ListChaptersCommand#listChapters()` — gets and lists all chapters in module level.
+
+For instance, the user wants to list all chapters available in `CS2113T` (module name), a detailed description of what happens is shown below:
+
+* Step 1: The user is currently in `CS2113T` level. 
+
+* Step 2: The user enters `list` command to list all chapters in `CS2113T` level. 
+
+The following sequence diagram shows how the list chapters feature works:
+![Sequence Diagram of List Chapters](UML/listchap_seq_diagram.png)
```
Comment (madbeez) [inline docs/DeveloperGuide.md]: image
Should this diagram include activation bars?

File: docs/DeveloperGuide.md [hunk 18]
```diff
@@ -1,34 +1,1015 @@
...
+Step 10: `GoModuleCommand#execute()` prints the performance message to the user via `Ui#showToUser` if the `ShowRateCommand#computePercentage()` is greater than int `0`.
+
+The following diagram shows how the show overall performance feature works:
+
+<p align="center">
+  <img src="UML/ShowRateCommand.png" width="800" alt="Sequence Diagram of show overall performance command"/>
+  <br/>Figure <>. Sequence Diagram of show overall performance command 
+</p>
+
+![Sequence Diagram of show overall performance command](UML/ShowRateCommand.png)
```
Comment (madbeez) [inline docs/DeveloperGuide.md]: Did you mean to include the same diagram twice?

File: docs/DeveloperGuide.md [hunk 19]
```diff
@@ -1,34 +1,1015 @@
...
+* `PreviewCommand#execute()` - Calls `PreviewCommand#loadAllDueChapters()`, before calling `PreviewCommand#setDueDueChapters()` for the current day, then one to six days after the current day, and prints the `DueChapter`s for each day individually
+
+On top of that, the following operations from `Storage` are used:
+* `Storage#checkChapterDeadline()` - Reads the deadline for each `Chapter`, prompts if they are corrupted, and adds a `DueChapter` formed with the `Chapter` into `ArrayList<DueChapter> allDueChapters`
+* `Storage#checkAllChaptersForDue()` - Obtains the name of every `Chapter` from the list of Modules passed to it and calls `Storage#checkChapterDeadline()` for each of them
+* `Storage#loadAllDueChapters()` - Obtains the names of every `Module` in the user database and calls `Storage#checkAllChaptersForDue()`
+
+The following sequence Diagrams illustrates how the Preview Upcoming Dues Process is executed:
```
Comment (madbeez) [inline docs/DeveloperGuide.md]: This section comes across as a wall of text, maybe you can improve the spacing to make it more readable?
